HARK/ConsumptionSaving/ConsRiskyContribModel.py

```diff
@@ -0,0 +1,1768 @@
+        (0) permanent income shock, (1) transitory income shock, (2) risky
+        asset return.
+    a : float
+        end-of-period risk-free assets.
+    nTil : float
+        end-of-period risky assets.
+    s : float
+        end-of-period income deduction share.
+    dvdm_next : 3D function
+        Next-period's marginal value of riskless assets.
+    dvdn_next : 3D function
+        Next-period's marginal value of risky assets.
+    dvds_next : 3D function
+        Next-period's marginal value of income-deduction share.
+    CRRA : float
+        Coefficient of relative risk aversion.
+    PermGroFac : float
+        Permanent income deterministic growth factor.
+    Rfree : float
+        Risk-free return factor.
+    DiscFac : float
+        Time-preference discount factor.
+    LivPrb : float
+        Survival probability.
+    v_next : 3D function, optional
+        Next-period's value function. The default is None.
+
+    Returns
+    -------
+    np.array
+        Array with end-of-period-value derivatives conditional on next
+        period's shocks. Order
+        (0) riskless assets, (1) risky assets, (2) income deduction share
+        Optionally, the level of end-of-period value is added in position (3).
+
+    """
+
+    temp_fac_A = utilityP(shocks[0] * PermGroFac, CRRA)
+    temp_fac_B = (shocks[0] * PermGroFac) ** (1.0 - CRRA)
+
+    # Find next-period asset balances
+    m_next = m_nrm_next(shocks, a, s, Rfree, PermGroFac)
+    n_next = n_nrm_next(shocks, nTil, s, PermGroFac)
+
+    # Interpolate next-period-value derivatives
+    dvdm_tp1 = dvdm_next(m_next, n_next, s)
+    dvdn_tp1 = dvdn_next(m_next, n_next, s)
+    if shocks[1] == 0:
+        dvds_tp1 = dvds_next(m_next, n_next, s)
+    else:
+        dvds_tp1 = shocks[1] * (dvdn_tp1 - dvdm_tp1) + dvds_next(m_next, n_next, s)
+
+    # Discount next-period-value derivatives to current period
+
+    # Liquid resources
+    end_of_prd_dvda = DiscFac * Rfree * LivPrb * temp_fac_A * dvdm_tp1
+    # Iliquid resources
+    end_of_prd_dvdn = DiscFac * shocks[2] * LivPrb * temp_fac_A * dvdn_tp1
+    # Contribution share
+    end_of_prd_dvds = DiscFac * LivPrb * temp_fac_B * dvds_tp1
+
+    # End of period value function, if needed
+    if v_next is not None:
+        end_of_prd_v = DiscFac * LivPrb * temp_fac_B * v_next(m_next, n_next, s)
+        return np.stack(
+            [end_of_prd_dvda, end_of_prd_dvdn, end_of_prd_dvds, end_of_prd_v]
+        )
+    else:
+        return np.stack([end_of_prd_dvda, end_of_prd_dvdn, end_of_prd_dvds])
+
+
+# %% RiskyContrib solvers
+
+# Consumption stage solver
+def solve_RiskyContrib_Cns(
+    solution_next,
+    ShockDstn,
+    LivPrb,
+    DiscFac,
+    CRRA,
+    Rfree,
+    PermGroFac,
+    BoroCnstArt,
+    aXtraGrid,
+    nNrmGrid,
+    mNrmGrid,
+    ShareGrid,
+    vFuncBool,
+    AdjustPrb,
+    DiscreteShareBool,
+    **unused_params
+):
+
+    # Make sure the individual is liquidity constrained.  Allowing a consumer to
```

Comment:
This single function is 250 lines long.

My suspicion from working with HARK is that much of this complexity is due to the hard-coding of mathematical operations that could and should be functionalized out into reusable subroutines.

I wonder if there is any way to shorten this and the other solution methods?

Reply:
That's a fair criticism. I'll see what I can do.

Incidentally, this is another pro of the "splitting into stages" strategy. Imagine if these were not 3 250-line functions but a single 750-line function.

Reply:
I've spun my wheels a bit on this and I don't see a reduction or partition that would result in functions that would be useful in other models. I could split the `solve_*` functions in parts but they would be specific to this model and require more arguments being juggled around, complexity would increase in my opinion.

You are right that a lot of complexity and raw code length comes from operations that could be abstracted. But I'm using a host of those things that we've generalized already e.g.
- The general n-dimensional `CRRA(Marg)ValueFunc` that are now in `HARK.interpolation` are a spin-off of this project.
- You'll be pleased to see that expectations are computed using your `calc_expectations` function.

I understand your point. But some of these things are hard to abstract. For instance, one might think of working towards a routine that "finds an optimization problem's solution given its first order condition". But in my model:
- The FOC is necessary and sufficient in the `Cns` stage. Yet, it is used in a very idiosyncratic way (EGM). This way of using it is worth it because EGM is the most efficient way one can solve a consumption problem if it is applicable.
- It is not necessary or sufficient in the `Reb` or `Sha` stage because of the possibility of solutions that are not interior (`Sha`) or kinks in the objective function (`Reb`). Still, the FOC contains information that allows you to find the solutions.

It's these idiosyncrasies that result in this "hard-coding" of solution methods. But it might be something we might have to live with. The opposite extreme alternative would be to use `scipy.optimize` to create a solver that would apply to any bellman equation.  But this solver would not have a chance of solving the kinds of models we are interested in. We are incorporating analytical insights of specific models to reduce numerical optimization demands.

Reply:
Thanks for thinking this through.
I was concerned when I mentioned it that it might be an unfair point to make; it's certainly not blocking the PR, which I intend to merge soon.

It looks like many of the lines are documentation and idiosyncrasies of the `black` formatting tool.

I think it should be possible to write a generic EGM solver but that is well beyond the scope of this issue.